### PR TITLE
[distributions] Always enable grad when calculating lazy_property

### LINKED ISSUE #7705: [distributions] rsample().detach() and sample() yields different gradients
I have a MultivariateNormal distribution with loc defined as output of neural net (given input) and diagonal covariance matrix with trainable parameters (but does not depending on some input). 
If I sample via `distr.rsample().detach()` and optimise sum of log_probs, `.backward()` provides correct gradients w.r.t. both loc and cov matrix params. But if I sample via `distr.sample()`, `.backward()` sets `None` gradients for cov matrix params.

Here is a minimal reproducing code:
```
import torch
from torch.distributions import MultivariateNormal

torch.manual_seed(124)
inp = torch.randn(3)
W = torch.randn((2, 3), requires_grad=True)
loc = W@inp
cov = torch.tensor([[1., 0.], [0., 2.]], requires_grad=True)
distr = MultivariateNormal(loc, cov)

y = distr.rsample().detach()  # difference here
loss = distr.log_prob(y)
loss.backward()
W.grad, cov.grad
```

yields 

```
(tensor([[-0.1991, -1.0776, -0.6339],
         [ 0.3455,  1.8700,  1.1000]]), 
 tensor([[-0.2678,  0.0000],
         [-0.8057,  0.4491]]))
```

while 

```
import torch
from torch.distributions import MultivariateNormal

torch.manual_seed(124)
inp = torch.randn(3)
W = torch.randn((2, 3), requires_grad=True)
loc = W@inp
cov = torch.tensor([[1., 0.], [0., 2.]], requires_grad=True)
distr = MultivariateNormal(loc, cov)

x = distr.sample()  # difference here
loss = distr.log_prob(x)
loss.backward()
W.grad, cov.grad
```
yields
```
(tensor([[-0.1991, -1.0776, -0.6339],
         [ 0.3455,  1.8700,  1.1000]]), 
 None)
```

I am observing this problem both on macOS and Ubuntu systems.

### PR DESCRIPTION
**Pull request by ssnl**
Fixes #7705 .

@apaszke @fritzo @alicanb

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -690,6 +690,23 @@ def test_enumerate_support_type(self):
...
+                return x + 1
+
+        def test():
+            x.grad = None
+            Dummy().y.backward()
+            self.assertEqual(x.grad, torch.ones(1))
+
+        test()
+        with torch.no_grad():
+            test()
```
Comment (apaszke) [inline test/test_distributions.py]: Can you also add a test with an actual distribution?

File: torch/distributions/utils.py [hunk 2]
```diff
@@ -182,6 +182,7 @@ def __init__(self, wrapped):
     def __get__(self, instance, obj_type=None):
         if instance is None:
             return self
-        value = self.wrapped(instance)
+        with torch.enable_grad():
```
Comment (apaszke) [inline torch/distributions/utils.py]: Meh, it's not the cleanest to have this as part of the `lazy_property` contract 😕 On the other hand moving it outside would make it quite error prone.

Comment (SsnL) [reply torch/distributions/utils.py]: I agree. But unless we want to have grad modes of two different priorities (user set has highest pri), I can't think of a clean solution.

Comment (t-vi) [reply torch/distributions/utils.py]: Maybe calculating remembering whether it has grad and re-calculating the thing when it's needed but not available?

Comment (apaszke) [reply torch/distributions/utils.py]: That won't work. The lazy property makes use of some special Python tricks that cause it to go through `__get__` exactly once, and later accesses to the attr will simply take the thing from the `__dict__` without even asking. They would need to become regular properties, and their contract would get even more complicated.

**Comment by apaszke** [comment]
cc @fritzo

File: torch/distributions/utils.py (hunk 2 above)
Comment (t-vi) [reply torch/distributions/utils.py]: If I understand the logic correctly, the logic is to use `__get__` unless there is an attribute of the same name. If you cache `value` in self and make recalculation and the `setattr` conditional on `is_grad_enabled`, you'd get a property that becomes lazy on the first invocation with grads. I'm not sure whether the overhead of calling `__get__` is that large in the grand scheme of things, but if it is regularly used with `no_grad`, I think the memory saving might be worth the compute.

**Comment by fritzo** [comment]
cc @neerajprad 

It seems a little forceful to always enable grad. One option that seems consistent to me is to inherit `grad_enabled` from its setting at initialization time
```py
class Distribution(object):
    def __init__(self, ...):
        ...
        self._grad_enabled = torch.autograd.get_grad_enabled()  # what is the actual syntax?

class lazy_property(object):
    ...
    def __get__(self, instance, obj_type=None):
        if instance is None:
            return self
        with torch.autograd.set_grad_enabled(instance._grad_enabled):
            value = self.wrapped(instance)
        setattr(instance, self.wrapped.__name__, value)
        return value
```
However it's unclear whether we should be clever or whether we should simply require strict usage: "grad_enabled should have a single value throughout the lifetime of a distribution object". The distributions are intended to be flyweight objects and to be easy to reconstruct in each grad_enabled context.

**Comment by apaszke** [comment]
I think the "constant grad mode" invariant is a bit too easy to get wrong. Forcing grad mode doesn't seem that bad really. If your distribution parameters don't require grad then it will be a no-op. Otherwise it's quite likely that you will be interested in differentiating those parts, and I think it's better to trade off some memory for compute in this case.

It's a hard problem, because you don't have any information about how the object will be used. You either need to drop the autograd history, hoping that someone who calls `.sample()` will never need it, or you need to just live with the fact that you might end up wasting some memory, because we temporarily enabled grad. Provided the distribution is short-lived the downside is somewhat irrelevant though, because the properties should get removed quickly. If we find out that this strategy doesn't work too well, we can always add an extra constructor parameter that lets you choose a strategy for caching those.

**Comment by SsnL** [comment]
@fritzo @apaszke @neerajprad have we reached consensus on whether we should merge this? :)

**Comment by fritzo** [comment]
This seems reasonable to me (I'm convinced by @apaszke).

**Review by apaszke — APPROVED** [review]
